File: backend/app/matcher/cut_match.py
from typing import Dict, Optional
from pathlib import Path


def cut_key(filename: str, prefix_cut: int, suffix_cut: int) -> str:
    """
    Extract matching key from filename using prefix/suffix cuts.
    Cuts apply to entire filename including extension.
    """
    if prefix_cut < 0 or suffix_cut < 0:
        raise ValueError("Cut values must be non-negative")
    if prefix_cut + suffix_cut > len(filename):
        raise ValueError("Total cut exceeds filename length")
    return filename[prefix_cut : len(filename) - suffix_cut]


class ImageMatcher:
    """Matches mdoc entries to image files using cut keys"""

    def __init__(self, image_dir: str, image_prefix_cut: int = 0, image_suffix_cut: int = 0):
        self.image_dir = Path(image_dir)
        self.image_prefix_cut = image_prefix_cut
        self.image_suffix_cut = image_suffix_cut
        self._cache: Dict[str, str] = {}

    def build_cache(self):
        """Scan image directory and build key -> path cache"""
        self._cache.clear()
        if not self.image_dir.exists():
            return

        for img_path in self.image_dir.rglob("*"):
            if img_path.is_file():
                # Cache both with and without extension
                key_with_ext = cut_key(img_path.name, self.image_prefix_cut, self.image_suffix_cut)
                key_no_ext = cut_key(img_path.stem, self.image_prefix_cut, self.image_suffix_cut)
                # Store with both keys, preferring the full name
                self._cache[key_with_ext] = str(img_path)
                self._cache[key_no_ext] = str(img_path)

    def match(self, filename: str) -> Optional[str]:
        """Match a filename to an image path"""
        # Try with extension first
        key = cut_key(filename, self.image_prefix_cut, self.image_suffix_cut)
        if key in self._cache:
            return self._cache[key]
        # Try without extension
        key_no_ext = cut_key(Path(filename).stem, self.image_prefix_cut, self.image_suffix_cut)
        return self._cache.get(key_no_ext)

File: backend/src/matcher/cut_match.rs
use std::path::Path;

/// Matches mdoc entries to image files by checking if the mdoc's filename
/// stem is contained within the mrc filename.
///
/// For example, an mdoc named `Position_1_13.mdoc` will match any mrc file
/// whose name contains `Position_1_13`, such as
/// `Position_1_13_006_-21.00_20260527_144344_EER.mrc`.
pub struct ImageMatcher {
    image_dir: String,
    /// List of (filename_stem, full_path) for all image files found
    files: Vec<(String, String)>,
}

impl ImageMatcher {
    pub fn new(image_dir: &str, _image_prefix_cut: usize, _image_suffix_cut: usize) -> Self {
        Self {
            image_dir: image_dir.to_string(),
            files: Vec::new(),
        }
    }

    /// Scan image directory and build a list of all image files.
    pub fn build_cache(&mut self) {
        self.files.clear();
        let dir = Path::new(&self.image_dir);
        if !dir.exists() {
            return;
        }

        let walker = walkdir::WalkDir::new(dir).max_depth(10);
        for entry in walker.into_iter().filter_map(|e| e.ok()) {
            if entry.file_type().is_file() {
                let path = entry.path().to_string_lossy().to_string();
                let name = entry.file_name().to_string_lossy().to_string();
                let stem = Path::new(&name)
                    .file_stem()
                    .map(|s| s.to_string_lossy().to_string())
                    .unwrap_or_default();
                if !stem.is_empty() {
                    self.files.push((stem, path));
                }
            }
        }
    }

    /// Match a filename to an image path.
    ///
    /// Checks if any known mrc stem contains the given filename as a substring.
    /// This handles the common case where mrc filenames embed the mdoc basename:
    ///   mdoc: Position_1_13.mdoc
    ///   mrc:  Position_1_13_006_-21.00_20260527_144344_EER.mrc
    pub fn match_filename(&self, filename: &str) -> Option<String> {
        if filename.is_empty() {
            return None;
        }

        for (stem, path) in &self.files {
            if stem.contains(filename) {
                return Some(path.clone());
            }
        }

        None
    }
}
